src/Xamarin.Android.Build.Tasks/Tasks/ProcessAssemblies.cs

```diff
@@ -46,82 +50,120 @@ public override bool RunTask ()
-								packageId.StartsWith ("Microsoft.Android.Runtime.");
-						}
-						if (hasMonoAndroidReference == null) {
-							hasMonoAndroidReference = MonoAndroidHelper.HasMonoAndroidReference (reader);
-						}
-						assembly.SetMetadata ("FrameworkAssembly", frameworkAssembly.ToString ());
-						assembly.SetMetadata ("HasMonoAndroidReference", hasMonoAndroidReference.ToString ());
+					using var pe = new PEReader (File.OpenRead (assembly.ItemSpec));
+					var reader = pe.GetMetadataReader ();
+					var module = reader.GetModuleDefinition ();
+					var mvid = reader.GetGuid (module.Mvid);
+					mvids.Add (mvid);
+
+					// Calculate %(FrameworkAssembly) and %(HasMonoAndroidReference) for the first
+					if (frameworkAssembly == null) {
+						frameworkAssembly = IsFrameworkAssembly (assembly);
+					}
+					if (hasMonoAndroidReference == null) {
+						hasMonoAndroidReference = MonoAndroidHelper.IsMonoAndroidAssembly (assembly) ||
+							MonoAndroidHelper.HasMonoAndroidReference (reader);
 					}
+					assembly.SetMetadata ("FrameworkAssembly", frameworkAssembly.ToString ());
+					assembly.SetMetadata ("HasMonoAndroidReference", hasMonoAndroidReference.ToString ());
 				}
 				// If we end up with more than 1 unique mvid, we need *all* assemblies
 				if (mvids.Count > 1) {
 					foreach (var assembly in group) {
-						var symbolPath = Path.ChangeExtension (assembly.ItemSpec, ".pdb");
-						if (!symbols.TryGetValue (symbolPath, out var symbol)) {
-							// Sometimes .pdb files are not included in @(ResolvedFileToPublish), so add them if they exist
-							if (File.Exists (symbolPath)) {
-								symbols [symbolPath] = symbol = new TaskItem (symbolPath);
-							}
-						}
+						var symbol = GetOrCreateSymbolItem(symbols, assembly);
 						SetDestinationSubDirectory (assembly, group.Key, symbol);
 						output.Add (assembly);
 					}
 				} else {
 					// Otherwise only include the first assembly
 					bool first = true;
 					foreach (var assembly in group) {
-						var symbolPath = Path.ChangeExtension (assembly.ItemSpec, ".pdb");
 						if (first) {
 							first = false;
-							if (!symbols.TryGetValue (symbolPath, out var symbol)) {
-								// Sometimes .pdb files are not included in @(ResolvedFileToPublish), so add them if they exist
-								if (File.Exists (symbolPath)) {
-									symbols [symbolPath] = symbol = new TaskItem (symbolPath);
-								}
-							}
+
+							var symbol = GetOrCreateSymbolItem (symbols, assembly);
 							symbol?.SetDestinationSubPath ();
 							assembly.SetDestinationSubPath ();
 							output.Add (assembly);
 						} else {
-							symbols.Remove (symbolPath);
+							symbols.Remove (Path.ChangeExtension (assembly.ItemSpec, ".pdb"));
 						}
 					}
 				}
 			}
+		}
 
-			OutputAssemblies = output.ToArray ();
-			ResolvedSymbols = symbols.Values.ToArray ();
+		static bool IsFrameworkAssembly (ITaskItem assembly)
+		{
+			string packageId = assembly.GetMetadata ("NuGetPackageId") ?? "";
```

Comment:
Documentation for [`ITaskItem.GetMetadata()`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.framework.itaskitem.getmetadata?view=msbuild-16-netcore) leaves a bit to be desired, but… will this return `null`, or `""`, when the specified metadata isn't found?

It might be "safer" to just use `string.IsNullOrEmpty()`.